internal/access/credentialrequest.go

```diff
@@ -0,0 +1,165 @@
+package access
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/infrahq/infra/api"
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/uid"
+)
+
+func CreateCredentialRequest(c *gin.Context, destination string) (*models.CredentialRequest, error) {
+	rCtx := GetRequestContext(c)
+	// does the user have a grant to this destination?
+	// ListGrants(c, )
+	grants, err := data.ListGrants(rCtx.DBTxn, data.ListGrantsOptions{
+		ByDestination: destination,
+		BySubject:     uid.NewIdentityPolymorphicID(rCtx.Authenticated.User.ID),
+		Pagination: &data.Pagination{
+			Limit: 1,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(grants) == 0 {
+		return nil, internal.ErrUnauthorized
+	}
+
+	dest, err := data.GetDestination(rCtx.DBTxn, data.GetDestinationOptions{ByName: destination})
+	if err != nil {
+		return nil, err
+	}
+
+	cr := &models.CredentialRequest{
+		ID:                 uid.New(),
+		OrganizationMember: models.OrganizationMember{OrganizationID: rCtx.Authenticated.User.OrganizationID},
+		ExpiresAt:          time.Now().Add(2 * time.Minute),
+		UserID:             rCtx.Authenticated.User.ID,
+		DestinationID:      dest.ID,
+	}
+	err = data.CreateCredentialRequest(rCtx.DBTxn, cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// commit the transaction so others can see the request
+	err = rCtx.DBTxn.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for the request to be filled; listen for the notify on the specific record req.ID
+	listener, err := data.ListenForNotify(c, rCtx.DataDB, data.ListenForNotifyOptions{
```

Comment:
Why do we wait for the request to be fulfilled rather than just finishing and letting the client do a long poll as a separate request? It seems possible to me that you may just want to request a credential then carry on until it is approved.